File: c/leet_code/200.c
#include <stdio.h>

/**
 * @file 200.c
 * @brief Contains the implementation of the number of islands problem.
 */

/**
 * @brief Performs a depth-first search (DFS) on the grid from a given cell.
 *
 * This function is used to mark the visited cells in the grid by changing their value to '0'.
 *
 * @param grid The 2D grid of cells.
 * @param i The row index of the current cell.
 * @param j The column index of the current cell.
 * @param rowSize The number of rows in the grid.
 * @param colSize The number of columns in the grid.
 * @return void
 */
void dfs(char **grid, int i, int j, int rowSize, int colSize) {
    // If the current cell is '0' (water), return.
    if (grid[i][j] == '0') return;

    // Mark the current cell as '0' (water).
    grid[i][j] = '0';

    // Perform DFS on the four adjacent cells if they are within the bounds of the grid.
    if (i + 1 < rowSize) dfs(grid, i + 1, j, rowSize, colSize);
    if (i - 1 >= 0) dfs(grid, i - 1, j, rowSize, colSize);
    if (j + 1 < colSize) dfs(grid, i, j + 1, rowSize, colSize);
    if (j - 1 >= 0) dfs(grid, i, j - 1, rowSize, colSize);
}

/**
 * @brief Counts the number of islands in the grid.
 *
 * An island is defined as a group of '1's (land) that are connected horizontally or vertically.
 * This function uses the DFS function to traverse the grid and count the number of islands.
 *
 * @param grid The 2D grid of cells.
 * @param gridSize The number of rows in the grid.
 * @param gridColSize The number of columns in the grid.
 * @return The number of islands in the grid.
 */
int numIslands(char **grid, int gridSize, int *gridColSize) {
    int ans = 0;

    // Traverse the grid.
    for (int i = 0; i < gridSize; ++i) {
        for (int j = 0; j < *gridColSize; ++j) {
            // If the current cell is '1' (land), perform DFS from here and increment the island count.
            if (grid[i][j] == '1') {
                ans++;
                dfs(grid, i, j, gridSize, *gridColSize);
            }
        }
    }

    return ans;
}

/**
 * @brief Example usage of the numIslands function.
 *
 * This function creates a grid and calls the numIslands function to count the number of islands in the grid.
 * The result is then printed to the standard output.
 *
 * @return void
 */
void exampleUsage() {
    // Create a 2D grid.
    char *grid[] = {"11000", "11000", "00100", "00011"};

    // Count the number of islands in the grid.
    int num_of_islands = numIslands(grid, 4, 5);

    // Print the result.
    printf("Number of islands: %d\n", num_of_islands);
}

int main() {
    exampleUsage();
    return 0;
}

File: python/leet_code/200.py
"""
This module provides a solution to the problem of counting the number of islands in a 2D grid map.
It reads the grid map from the standard input, counts the number of islands, and writes the results to a file.
"""

from json import loads as l


def dfs(g: list[list[str]], i: int, j: int) -> list[list[str]]:
    """
    This function performs a depth-first search (DFS) on the grid map to find all the cells that belong to the same island.

    Args:
        g (List[List[str]]): A 2D grid map of '1's (land) and '0's (water).
        i (int): The row index of the current cell.
        j (int): The column index of the current cell.

    Returns:
        List[List[str]]: The updated grid map after marking the visited cells.

    Example:
        >>> dfs(
        ...     [
        ...         ['1', '1', '1', '1', '0'],
        ...         ['1', '1', '0', '1', '0'],
        ...         ['1', '1', '0', '0', '0'],
        ...         ['0', '0', '0', '0', '0'],
        ...     ],
        ...     0,
        ...     0,
        ... )
        [['0', '0', '0', '0', '0'], ['0', '0', '0', '0', '0'], ['0', '0', '0', '0', '0'], ['0', '0', '0', '0', '0']]

    Doctest:
        >>> dfs(
        ...     [
        ...         ['1', '1', '0', '0', '0'],
        ...         ['1', '1', '0', '0', '0'],
        ...         ['0', '0', '1', '0', '0'],
        ...         ['0', '0', '0', '1', '1'],
        ...     ],
        ...     0,
        ...     0,
        ... )
        [['0', '0', '0', '0', '0'], ['0', '0', '0', '0', '0'], ['0', '0', '1', '0', '0'], ['0', '0', '0', '1', '1']]
        >>> dfs(
        ...     [
        ...         ['1', '1', '0', '0', '0'],
        ...         ['1', '1', '0', '0', '0'],
        ...         ['0', '0', '1', '0', '0'],
        ...         ['0', '0', '0', '1', '1'],
        ...     ],
        ...     2,
        ...     2,
        ... )
        [['1', '1', '0', '0', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '0', '0', '0'], ['0', '0', '0', '1', '1']]
        >>> dfs(
        ...     [
        ...         ['1', '1', '0', '0', '0'],
        ...         ['1', '1', '0', '0', '0'],
        ...         ['0', '0', '1', '0', '0'],
        ...         ['0', '0', '0', '1', '1'],
        ...     ],
        ...     3,
        ...     3,
        ... )
        [['1', '1', '0', '0', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '1', '0', '0'], ['0', '0', '0', '0', '0']]
        >>> dfs(
        ...     [
        ...         ['1', '1', '0', '0', '0'],
        ...         ['1', '1', '0', '0', '0'],
        ...         ['0', '0', '1', '0', '0'],
        ...         ['0', '0', '0', '1', '1'],
        ...     ],
        ...     3,
        ...     4,
        ... )
        [['1', '1', '0', '0', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '1', '0', '0'], ['0', '0', '0', '1', '0']]
        >>> dfs(
        ...     [
        ...         ['1', '1', '0', '0', '0'],
        ...         ['1', '1', '0', '0', '0'],
        ...         ['0', '0', '1', '0', '0'],
        ...         ['0', '0', '0', '1', '1'],
        ...     ],
        ...     0,
        ...     4,
        ... )
        [['1', '1', '0', '0', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '1', '0', '0'], ['0', '0', '0', '1', '1']]
    """
    g[i][j] = '0'
    if i + 1 < len(g) and g[i + 1][j] == '1':
        g = dfs(g, i + 1, j)
    if j + 1 < len(g[i]) and g[i][j + 1] == '1':
        g = dfs(g, i, j + 1)
    if i - 1 >= 0 and g[i - 1][j] == '1':
        g = dfs(g, i - 1, j)
    if j - 1 >= 0 and g[i][j - 1] == '1':
        g = dfs(g, i, j - 1)
    return g


with open('user.out', 'w') as file:
    for g in stdin:
        g = l(g)
        if not g or not g[0]:
            file.write('0' + '\n')
        else:
            c = 0
            for i in range(len(g)):
                for j in range(len(g[i])):
                    if g[i][j] == '1':
                        c += 1
                        g = dfs(g, i, j)
            file.write(str(c) + '\n')
exit()
